extensions/ffmpeg/src/main/jni/ffmpeg_jni.cc

```diff
@@ -365,3 +386,342 @@ void releaseContext(AVCodecContext *context) {
   avcodec_free_context(&context);
 }
 
+
+// video
+
+// YUV plane indices.
+const int kPlaneY = 0;
+const int kPlaneU = 1;
+const int kPlaneV = 2;
+const int kMaxPlanes = 3;
+
+// Android YUV format. See:
+// https://developer.android.com/reference/android/graphics/ImageFormat.html#YV12.
+const int kImageFormatYV12 = 0x32315659;
+
+struct JniContext {
+  ~JniContext() {
+    if (native_window) {
+      ANativeWindow_release(native_window);
+    }
+  }
+
+  bool MaybeAcquireNativeWindow(JNIEnv *env, jobject new_surface) {
+    if (surface == new_surface) {
+      return true;
+    }
+    if (native_window) {
+      ANativeWindow_release(native_window);
+    }
+    native_window_width = 0;
+    native_window_height = 0;
+    native_window = ANativeWindow_fromSurface(env, new_surface);
+    if (native_window == nullptr) {
+      LOGE("kJniStatusANativeWindowError");
+      surface = nullptr;
+      return false;
+    }
+    surface = new_surface;
+    return true;
+  }
+
+  jfieldID data_field;
+  jfieldID yuvPlanes_field;
+  jfieldID yuvStrides_field;
+  jmethodID init_for_private_frame_method;
+  jmethodID init_for_yuv_frame_method;
+  jmethodID init_method;
+
+  AVCodecContext *codecContext;
+
+  ANativeWindow *native_window = nullptr;
+  jobject surface = nullptr;
+  int native_window_width = 0;
+  int native_window_height = 0;
+};
+
+void CopyPlane(const uint8_t *source, int source_stride, uint8_t *destination,
+               int destination_stride, int width, int height) {
+  while (height--) {
+    std::memcpy(destination, source, width);
```

Comment:
Sorry if this is the wrong place to put this but,

It seems like CopyPlane needs to take the strides into account when copying data. Something like 
`std::memcpy(destination, source, std::min(std::min(width, source_stride), destination_stride));` 
prevents reading past the end of `source` or writing past the end of `destination`. 

I'm pretty sure this is responsible for the issue reported here https://github.com/google/ExoPlayer/issues/9459